File: ocaml/21-30/25.ml
(* Generate a random permutation of the elements of a list. (easy)
*)

let shuffle a = for n = ((Array.length a) - 1) downto 0 do 
  Random.init(43799); 
  let k = Random.int (n + 1) in 
  let temp = a.(n) in 
  a.(n) <- a.(k); 
  a.(k) <- temp 
done 

let permutation xs = let a = Array.of_list xs in
  shuffle a;
  Array.to_list a

let _ = assert ((permutation ["a"; "b"; "c"; "d"; "e"; "f"]) = ["c"; "f"; "a"; "b"; "e"; "d"])

File: prolog/21-30/25.pl
/*
Generate a random permutation of the elements of a list.
    Example:
    ?- rnd_permu([a,b,c,d,e,f],L).
    L = [b,a,d,c,e,f]
*/

:- include('23').

rnd_permute(Xs, Ys) :- 
  length(Xs, L),
  rnd_select(Xs, L, Ys).
